src/FsAutoComplete.Core/Decompiler.fs

```diff
@@ -0,0 +1,233 @@
+        propertiesPosition:IReadOnlyDictionary<PropertyDef, Position> *
+        eventsPosition:IReadOnlyDictionary<EventDef, Position> *
+        typesPosition:IReadOnlyDictionary<TypeDef, Position>
+
+type OutputPositions with
+    member this.TryFindMethod def =
+        let (OutputPositions (methodsPosition=methodsPosition)) = this
+        methodsPosition |> tryGetValue def
+
+    member this.TryFindField def =
+        let (OutputPositions (fieldsPosition=fieldsPosition)) = this
+        fieldsPosition |> tryGetValue def
+
+    member this.TryFindProperty def =
+        let (OutputPositions (propertiesPosition=propertiesPosition)) = this
+        propertiesPosition |> tryGetValue def
+
+    member this.TryFindEvent def =
+        let (OutputPositions (eventsPosition=eventsPosition)) = this
+        eventsPosition |> tryGetValue def
+
+    member this.TryFindType def =
+        let (OutputPositions (typesPosition=typesPosition)) = this
+        typesPosition |> tryGetValue def
+
+
+type TextWriterWithPositions(writer: TextWriter) =
+    inherit TextWriterDecompilerOutput(writer)
+
+    let mutable line = 1
+    let mutable lineStartAtPosition = 0
+    let methodsPosition = System.Collections.Generic.Dictionary<MethodDef, Position>()
+    let fieldsPosition = System.Collections.Generic.Dictionary<FieldDef, Position>()
+    let propertiesPosition = System.Collections.Generic.Dictionary<PropertyDef, Position>()
+    let eventsPosition = System.Collections.Generic.Dictionary<EventDef, Position>()
+    let typesPosition = System.Collections.Generic.Dictionary<TypeDef, Position>()
+
+    let storeDefinitionIfNeeded nextPosition (reference:obj) =
+        let position = Position(line, nextPosition - lineStartAtPosition)
+        match reference with
+        | :? MethodDef as def -> methodsPosition.[def] <- position
+        | :? FieldDef as def -> fieldsPosition.[def] <- position
+        | :? PropertyDef as def -> propertiesPosition.[def] <- position
+        | :? EventDef as def -> eventsPosition.[def] <- position
+        | :? TypeDef as def -> typesPosition.[def] <- position
+        | _ -> ()
+
+    override this.WriteLine() =
+        base.WriteLine()
+        line <- line + 1
+        lineStartAtPosition <- this.Length
+
+    override this.Write(text:string, reference:obj, flags:DecompilerReferenceFlags, color:obj ) =
+        storeDefinitionIfNeeded this.NextPosition reference
+
+        base.Write(text, reference, flags, color)
+
+    override this.Write(text:string, index, length, reference:obj, flags:DecompilerReferenceFlags, color:obj ) =
+        storeDefinitionIfNeeded this.NextPosition reference
+
+        base.Write(text, index, length, reference, flags, color)
+
+    member __.Positions
+        with get () = OutputPositions (  methodsPosition :> IReadOnlyDictionary<MethodDef, Position>,
+                                         fieldsPosition :> IReadOnlyDictionary<FieldDef, Position>,

```

Comment:
returning the casted Dictionary, may generate a race condition, because the same object of type`Dictionary` is returned and that is not thread safe. (if something change the dictionary, while some other thread enumerate it, will break)

because yuo just return the position in a point of time from `Positions` you can maybe use in `OutputPositions` the `Map` instead of `ReadOnlyDictionary`, like `MethodsPosition: Map<MethodDef, Position>`.
This will allow to cleanup removing the `tryGetValue`, because you can use builtin `Map.tryFind` (same behaviour, different type)